gobblin-utility/src/main/java/gobblin/util/request_allocation/ConcurrentBoundedPriorityIterable.java

```diff
@@ -0,0 +1,150 @@
+ * priority to low priority.
+ *
+ * Note: as with a priority queue, e1 < e2 means that e1 is higher priority than e2.
+ */
+@Slf4j
+public class ConcurrentBoundedPriorityIterable<T> implements Iterable<AllocatedRequestsBase.TAndRequirement<T>> {
+
+  private final TreeSet<AllocatedRequestsBase.TAndRequirement<T>> elements;
+  private final int dimensions;
+  @Getter
+  private final Comparator<? super T> comparator;
+  private final Comparator<AllocatedRequestsBase.TAndRequirement<T>> allDifferentComparator;
+  private final ResourceEstimator<T> estimator;
+  private final ResourcePool resourcePool;
+  private final ResourceRequirement currentRequirement;
+  private boolean rejectedElement = false;
+
+  public ConcurrentBoundedPriorityIterable(final Comparator<? super T> prioritizer, ResourceEstimator<T> resourceEstimator, ResourcePool pool) {
+
+    this.estimator = resourceEstimator;
+    this.resourcePool = pool;
+    this.dimensions = this.resourcePool.getNumDimensions();
+    this.comparator = prioritizer;
+    this.allDifferentComparator = new AllDifferentComparator();
+    this.elements = new TreeSet<>(this.allDifferentComparator);
+
+    this.currentRequirement = this.resourcePool.getResourceRequirementBuilder().zero().build();
+  }
+
+  /**
+   * This is the actual {@link Comparator} used in the {@link TreeSet}. Since {@link TreeSet}s use the provided
+   * {@link Comparator} to determine equality, we must force elements with the same priority to be different according
+   * to the {@link TreeSet}.
+   */
+  private class AllDifferentComparator implements Comparator<AllocatedRequestsBase.TAndRequirement<T>> {
+    @Override
+    public int compare(AllocatedRequestsBase.TAndRequirement<T> t1, AllocatedRequestsBase.TAndRequirement<T> t2) {
+      int providedComparison = ConcurrentBoundedPriorityIterable.this.comparator.compare(t1.getT(), t2.getT());
+      if (providedComparison != 0) {
+        return providedComparison;
+      }
+      return Long.compare(t1.getId(), t2.getId());
+    }
+  }
+
+  /**
+   * Offer an element to the container.
+   * @return true if the element was added, false if there was no space and we could not evict any elements to make it fit.
+   *         Note that the element may get evicted by future offers, so a return of true is not a guarantee that the
+   *         element will be present at any time in the future.
+   */
+  public boolean add(T t) {
+    boolean addedWorkunits = addImpl(t);
+    if (!addedWorkunits) {
+      this.rejectedElement = true;
+    }
+    return addedWorkunits;
+  }
+
+  private synchronized boolean addImpl(T t) {
+    AllocatedRequestsBase.TAndRequirement<T> newElement = new AllocatedRequestsBase.TAndRequirement<>(t, this.estimator.estimateRequirement(t, this.resourcePool));
+
+    if (this.resourcePool.exceedsHardBound(newElement.getResourceRequirement(), false)) {
+      // item does not fit even in empty pool
+      return false;
```

Comment:
We need logging or some other way to track when items are rejected. For example, we risk having an item being infinitely rejected if it was too big for the current resource limits. Currently, it is not easy to figure out why an item gets rejected.